### PR TITLE
Upgrade LLVM and Clang to current ToT

### PR DESCRIPTION
**Pull request by alan-baker**
* Fix uses of getOrInsertFunction
  * return type changed
* Fix uses of raw_fd_ostream
  * enums changed
* Fix memcpy replacement
  * alignments are now src and dest params
* Add default verification prefix
  * needs to know what string to use with -verify
* Test updates
  * struct inserts needs new targeted tests
  * pointer selection needs new targeted tests

### CODE REVIEW

File: test/Structs/extract_value.cl [hunk 1]
```diff
@@ -5,85 +5,49 @@
...
-// CHECK-DAG: %[[FLOAT_DYNAMIC_ARRAY_TYPE_ID]] = OpTypeRuntimeArray %[[FLOAT_TYPE_ID]]
-// CHECK-DAG: %[[ARG_STRUCT_TYPE_ID]] = OpTypeStruct %[[FLOAT_DYNAMIC_ARRAY_TYPE_ID]]
-// CHECK-DAG: %[[ARG_POINTER_TYPE_ID:[a-zA-Z0-9_]*]] = OpTypePointer StorageBuffer %[[ARG_STRUCT_TYPE_ID]]
-// CHECK-DAG: %[[UINT_TYPE_ID:[a-zA-Z0-9_]*]] = OpTypeInt 32 0
-// CHECK-DAG: %[[VOID_TYPE_ID:[a-zA-Z0-9_]*]] = OpTypeVoid
-// CHECK-DAG: %[[FOO_TYPE_ID:[a-zA-Z0-9_]*]] = OpTypeFunction %[[VOID_TYPE_ID]]
-// CHECK-DAG: %[[THING_TYPE_ID]] = OpTypeStruct %[[FLOAT_TYPE_ID]] %[[FLOAT_TYPE_ID]]
-
-
+// CHECK: %[[FLOAT_TYPE_ID:[a-zA-Z0-9_]*]] = OpTypeFloat 32
```
Comment (alan-baker) [inline test/Structs/extract_value.cl]: This test changed because LLVM had optimized to a constant store.

File: test/Structs/insert_value.cl [hunk 2]
```diff
@@ -5,90 +5,56 @@
...
-// CHECK-DAG: %[[ARG_STRUCT_TYPE_ID]] = OpTypeStruct %[[FLOAT_DYNAMIC_ARRAY_TYPE_ID]]
-// CHECK-DAG: %[[ARG_POINTER_TYPE_ID:[a-zA-Z0-9_]*]] = OpTypePointer StorageBuffer %[[ARG_STRUCT_TYPE_ID]]
-// CHECK-DAG: %[[UINT_TYPE_ID:[a-zA-Z0-9_]*]] = OpTypeInt 32 0
-// CHECK-DAG: %[[VOID_TYPE_ID:[a-zA-Z0-9_]*]] = OpTypeVoid
-// CHECK-DAG: %[[FOO_TYPE_ID:[a-zA-Z0-9_]*]] = OpTypeFunction %[[VOID_TYPE_ID]]
-// CHECK-DAG: %[[THING_TYPE_ID]] = OpTypeStruct %[[FLOAT_TYPE_ID]] %[[FLOAT_TYPE_ID]]
-// CHECK-DAG: %[[A_AND_B_TYPE_ID:[a-zA-Z0-9_]*]] = OpTypeFunction %[[THING_TYPE_ID]]
-
-
+// CHECK: %[[FLOAT_TYPE_ID:[a-zA-Z0-9_]*]] = OpTypeFloat 32
```
Comment (alan-baker) [inline test/Structs/insert_value.cl]: Same here

File: test/for.cl [hunk 3]
```diff
@@ -5,78 +5,30 @@
...
-// CHECK: OpFunctionEnd
-
 void kernel __attribute__((reqd_work_group_size(1, 1, 1))) foo(global uint* a, uint b)
 {
   for (uint i = 0; i < b; i++)
   {
     a[i]++;
   }
 }
+
```
Comment (alan-baker) [inline test/for.cl]: Bounds were different so I simplified the checks.

File: test/widen_switch_condition.cl [hunk 4]
```diff
@@ -48,182 +48,7 @@ kernel void foo(global float *out, float x, float y) {
...
-// CHECK:  [[_float:%[0-9a-zA-Z_]+]] = OpTypeFloat 32
-// CHECK:  [[__runtimearr_float]] = OpTypeRuntimeArray [[_float]]
-// CHECK:  [[__struct_3]] = OpTypeStruct [[__runtimearr_float]]
-// CHECK:  [[__ptr_StorageBuffer__struct_3:%[0-9a-zA-Z_]+]] = OpTypePointer StorageBuffer [[__struct_3]]
-// CHECK:  [[__struct_5]] = OpTypeStruct [[_float]]
-// CHECK:  [[__ptr_StorageBuffer__struct_5:%[0-9a-zA-Z_]+]] = OpTypePointer StorageBuffer [[__struct_5]]
-// CHECK:  [[_void:%[0-9a-zA-Z_]+]] = OpTypeVoid
-// CHECK:  [[_8:%[0-9a-zA-Z_]+]] = OpTypeFunction [[_void]]
-// CHECK:  [[_uint:%[0-9a-zA-Z_]+]] = OpTypeInt 32 0
+// CHECK-NOT: OpTypeInt
```
Comment (alan-baker) [inline test/widen_switch_condition.cl]: This seemed to be key the check occurring

File: deps.json [hunk 5]
```diff
@@ -6,15 +6,15 @@
       "subrepo" : "llvm-mirror/llvm",
       "branch" : "master",
       "subdir" : "third_party/llvm",
-      "commit" : "82f73ee5b37a2a4cc1bdad02bebaaaba71b65400"
+      "commit" : "26882c9d258b62748a7266207513a06990c8decc"
```
Comment (alan-baker) [inline deps.json]: This gets us in LLVM and Clang 9

Comment (dnovillo) [reply deps.json]: I'm going to have to make sure that `clspv-opt` still works with this.  Do you know if LLVM 9 removed support for the old pass manager to `opt`?

Comment (alan-baker) [reply deps.json]: We still use it in clspv, so it works.

**Comment by kpet** [comment]
If you have to iterate further on this one, you may want to get rid of the TODO I added in `ClusterPodKernelArgumentsPass`.

**Comment by alan-baker** [comment]
> If you have to iterate further on this one, you may want to get rid of the TODO I added in `ClusterPodKernelArgumentsPass`.

I've read the TODO, but I'm not certain what it entails. Is it just pushing everything in Attributes.getFnAttributes() into AttrBuildInfo?

File: lib/ReplaceLLVMIntrinsicsPass.cpp [hunk 6]
```diff
@@ -238,8 +239,7 @@ bool ReplaceLLVMIntrinsicsPass::replaceMemcpy(Module &M) {
 
           // Check that the alignment is a constant integer.
           assert(isa<ConstantInt>(CI->getArgOperand(3)));
```
Comment (dneto0) [inline lib/ReplaceLLVMIntrinsicsPass.cpp]: The assert on line 241 can be removed, as well as the comment before it.

File: test/VariablePointers/select_ssbo_same_buffer.cl [hunk 7]
```diff
@@ -5,7 +5,8 @@
 // RUN: FileCheck %s < %t2.spvasm
 // RUN: spirv-val --target-env vulkan1.0 %t.spv
 
-// Same object selection only requires VariablePointersStorageBuffer.
+// LLVM optimizes the selection to be between 1 and 2 and not pointers, so no
```
Comment (dneto0) [inline test/VariablePointers/select_ssbo_same_buffer.cl]: Clever compiler

File: test/hack_inserts_undef.cl [hunk 8]
```diff
@@ -72,7 +72,7 @@ kernel void foo(global S* data, float f) {
 // CHECK-DAG: [[_uint_3:%[0-9a-zA-Z_]+]] = OpConstant [[_uint]] 3
 // CHECK-DAG: [[_float_2:%[0-9a-zA-Z_]+]] = OpConstant [[_float]] 2
 // CHECK-DAG: [[_float_1:%[0-9a-zA-Z_]+]] = OpConstant [[_float]] 1
-// CHECK-DAG: [[_float_0:%[0-9a-zA-Z_]+]] = OpConstant [[_float]] 0
+// CHECK-DAG: [[_undef:%[0-9a-zA-Z_]+]] = OpUndef [[_float]]
```
Comment (dneto0) [inline test/hack_inserts_undef.cl]: Some drivers will also require -hack-undef, which will change this to the constant 0 anyway.

**Comment by kpet** [comment]
>I've read the TODO, but I'm not certain what it entails. Is it just pushing everything in Attributes.getFnAttributes() into AttrBuildInfo?

Yes, instead of creating the list from `AttrBuildInfo` and then re-creating another one adding the function attributes. It's the more natural way of writing this and what I did initially but was broken in the LLVM version previously used. It's a really minor cleanup. I added the TODO mainly to avoid others the pain of debugging what I'd already debugged.

File: lib/ClusterPodKernelArgumentsPass.cpp [hunk 9]
```diff
@@ -201,16 +201,13 @@ bool ClusterPodKernelArgumentsPass::runOnModule(Module &M) {
...
-    // differently since the version of LLVM used by clspv has a bug in get().
-    // TODO(kpet): Add function attributes to AttrBuildInfo when updating
-    //             LLVM past r330136.
-    auto newAttributes = AttributeList::get(M.getContext(), AttrBuildInfo);
+    // And finally function attributes.
     if (Attributes.hasAttributes(AttributeList::FunctionIndex)) {
-      auto attrs = AttrBuilder(Attributes.getFnAttributes());
-      auto idx = AttributeList::FunctionIndex;
-      newAttributes = newAttributes.addAttributes(M.getContext(), idx, attrs);
+      auto idx = Attributes+List::FunctionIndex;
```
Comment (alan-baker) [inline lib/ClusterPodKernelArgumentsPass.cpp]: @kpet PTAL

Comment (kpet) [reply lib/ClusterPodKernelArgumentsPass.cpp]: LGTM, thanks!

**Pull request merged**